File: src/mcp_server.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import {
  CallToolResult,
  GetPromptRequest,
  GetPromptRequestSchema,
  GetPromptResult,
  ListPromptsRequest,
  ListPromptsRequestSchema,
  ListPromptsResult,
} from '@modelcontextprotocol/sdk/types.js';
import { ChatPromptClient, Langfuse } from 'langfuse';
import { z } from 'zod';

import { extractVariables } from './utils.js';

// Requires Environment Variables
const langfuse = new Langfuse();

async function getPromptHandler(
  request: GetPromptRequest,
): Promise<GetPromptResult> {
  console.log('Handing get prompt request');

  const promptName: string = request.params.name;
  const args = request.params.arguments ?? {};

  try {
    // Initialize Langfuse client and fetch the prompt by name.
    let compiledTextPrompt: string | undefined;
    let compiledChatPrompt: ChatPromptClient['prompt'] | undefined; // Langfuse chat prompt type

    try {
      // try chat prompt type first
      const prompt = await langfuse.getPrompt(promptName, undefined, {
        type: 'chat',
      });
      if (prompt.type !== 'chat') {
        throw new Error(`Prompt '${promptName}' is not a chat prompt`);
      }
      compiledChatPrompt = prompt.compile(args);
    } catch {
      // fallback to text prompt type
      const prompt = await langfuse.getPrompt(promptName, undefined, {
        type: 'text',
      });
      compiledTextPrompt = prompt.compile(args);
    }

    if (compiledChatPrompt) {
      const result: GetPromptResult = {
        messages: compiledChatPrompt.map((msg) => ({
          content: {
            text: msg.content,
            type: 'text',
          },
          role: ['ai', 'assistant'].includes(msg.role) ? 'assistant' : 'user',
        })),
      };
      return result;
    } else if (compiledTextPrompt) {
      const result: GetPromptResult = {
        messages: [
          {
            content: { text: compiledTextPrompt, type: 'text' },
            role: 'user',
          },
        ],
      };
      return result;
    } else {
      throw new Error(`Failed to get prompt for '${promptName}'`);
    }
  } catch (err: unknown) {
    const error = err as Error;

    throw new Error(
      `Failed to get prompt for '${promptName}': ${error.message}`,
    );
  }
}

async function listPromptsHandler(
  request: ListPromptsRequest,
): Promise<ListPromptsResult> {
  console.log('Handing list prompt request');

  try {
    const cursor = request.params?.cursor;
    const page = cursor ? Number(cursor) : 1;
    if (cursor !== undefined && isNaN(page)) {
      throw new Error('Cursor must be a valid number');
    }

    const res = await langfuse.api.promptsList({
      label: 'production',
      limit: 100,
      page,
    });

    const resPrompts: ListPromptsResult['prompts'] = await Promise.all(
      res.data.map(async (i) => {
        const prompt = await langfuse.getPrompt(i.name, undefined, {
          cacheTtlSeconds: 0,
        });
        const variables = extractVariables(JSON.stringify(prompt.prompt));
        return {
          arguments: variables.map((v) => ({
            name: v,
            required: false,
          })),
          name: i.name,
        };
      }),
    );

    return {
      nextCursor:
        res.meta.totalPages > page ? (page + 1).toString() : undefined,
      prompts: resPrompts,
    };
  } catch (error) {
    console.error('Error fetching prompts:', error);
    throw new Error('Failed to fetch prompts');
  }
}

export const getMcpServer = (): McpServer => {
  const server = new McpServer(
    {
      name: 'langfuse-prompts',
      version: '1.0.0',
    },
    {
      capabilities: {
        prompts: {},
      },
    },
  );

  // Register handlers
  server.server.setRequestHandler(ListPromptsRequestSchema, listPromptsHandler);
  server.server.setRequestHandler(GetPromptRequestSchema, getPromptHandler);

  // Tools for compatibility
  server.tool(
    'get-prompts',
    'Get prompts that are stored in Langfuse',
    {
      cursor: z
        .string()
        .optional()
        .describe('Cursor to paginate through prompts'),
    },
    async (args) => {
      try {
        const res = await listPromptsHandler({
          method: 'prompts/list',
          params: {
            cursor: args.cursor,
          },
        });

        const parsedRes: CallToolResult = {
          content: res.prompts.map((p) => ({
            text: JSON.stringify(p),
            type: 'text',
          })),
        };

        return parsedRes;
      } catch (err: unknown) {
        const error = err as Error;
        return {
          content: [{ text: 'Error: ' + error.message, type: 'text' }],
          isError: true,
        };
      }
    },
  );

  server.tool(
    'get-prompt',
    'Get a prompt that is stored in Langfuse',
    {
      arguments: z
        .record(z.string())
        .optional()
        .describe(
          'Arguments with prompt variables to pass to the prompt template, json object, e.g. {"<name>":"<value>"}',
        ),
      name: z
        .string()
        .describe(
          'Name of the prompt to retrieve, use get-prompts to get a list of prompts',
        ),
    },
    async (args) => {
      try {
        const res = await getPromptHandler({
          method: 'prompts/get',
          params: {
            arguments: args.arguments,
            name: args.name,
          },
        });

        const parsedRes: CallToolResult = {
          content: [
            {
              text: JSON.stringify(res),
              type: 'text',
            },
          ],
        };

        return parsedRes;
      } catch (err: unknown) {
        const error = err as Error;

        return {
          content: [{ text: 'Error: ' + error.message, type: 'text' }],
          isError: true,
        };
      }
    },
  );

  return server;
};

File: src/mcp_server.py
import json
from typing import Any, Dict, List, Optional

from fastmcp import FastMCP, Handle, Message
from fastmcp.data_format import Text
from fastmcp.prompts import GetPrompt, GetPromptResponse, ListPrompts, ListPromptsResponse
from langfuse import Langfuse
from langfuse.model import ChatPromptClient

from .config import settings
from .utils import extract_variables

# Initialize Langfuse client from environment variables
langfuse = Langfuse(
    public_key=settings.LANGFUSE_PUBLIC_KEY,
    secret_key=settings.LANGFUSE_SECRET_KEY,
    base_url=settings.LANGFUSE_BASEURL,
)


def get_mcp_server() -> FastMCP:
    """
    Creates and configures the FastMCP server with prompt handlers and tools.
    """
    mcp = FastMCP(
        name="langfuse-prompts",
        version="1.0.0",
    )

    @mcp.prompt(name="prompts/list")
    async def list_prompts_handler(request: ListPrompts) -> ListPromptsResponse:
        """
        Handles the prompts/list request.
        Fetches prompts from the Langfuse API.
        """
        print("Handling list prompt request")
        try:
            page = int(request.cursor) if request.cursor else 1
            res = langfuse.api.prompts.list(
                label="production",
                limit=100,
                page=page,
            )

            res_prompts = []
            for i in res.data:
                # Inefficiently fetch each prompt to get variables, matching original logic
                prompt = langfuse.get_prompt(i.name, cache_ttl_seconds=0)
                variables = extract_variables(json.dumps(prompt.prompt))
                res_prompts.append(
                    {
                        "name": i.name,
                        "arguments": [{"name": v, "required": False} for v in variables],
                    }
                )

            next_cursor = str(page + 1) if res.meta.total_pages > page else None

            return ListPromptsResponse(
                prompts=res_prompts,
                next_cursor=next_cursor,
            )
        except Exception as e:
            print(f"Error fetching prompts: {e}")
            raise

    @mcp.prompt(name="prompts/get")
    async def get_prompt_handler(request: GetPrompt) -> GetPromptResponse:
        """
        Handles the prompts/get request.
        Fetches a specific prompt and compiles it.
        """
        print(f"Handling get prompt request for '{request.name}'")
        try:
            compiled_prompt: Optional[List[Dict[str, Any]]] = None

            try:
                # Try chat prompt first
                prompt = langfuse.get_prompt(request.name, type="chat")
                if isinstance(prompt, ChatPromptClient):
                    compiled = prompt.compile(**request.arguments)
                    compiled_prompt = [
                        {
                            "role": "assistant" if msg.role in ["ai", "assistant"] else "user",
                            "content": [{"type": "text", "text": msg.content}],
                        }
                        for msg in compiled
                    ]
            except Exception:
                # Fallback to text prompt
                prompt = langfuse.get_prompt(request.name, type="text")
                compiled = prompt.compile(**request.arguments)
                compiled_prompt = [
                    {
                        "role": "user",
                        "content": [{"type": "text", "text": compiled}],
                    }
                ]

            if compiled_prompt is None:
                raise ValueError(f"Failed to get prompt for '{request.name}'")

            return GetPromptResponse(messages=compiled_prompt)

        except Exception as e:
            print(f"Error getting prompt '{request.name}': {e}")
            raise

    @mcp.tool()
    async def get_prompts(
        handle: Handle,
        cursor: Optional[str] = None,
    ) -> List[Text]:
        """
        Get prompts that are stored in Langfuse.

        :param cursor: Cursor to paginate through prompts.
        """
        try:
            res = await list_prompts_handler(ListPrompts(cursor=cursor))
            return [Text(json.dumps(p)) for p in res.prompts]
        except Exception as e:
            await handle.error(f"Error: {e}")
            return []

    @mcp.tool()
    async def get_prompt(
        handle: Handle,
        name: str,
        arguments: Optional[str] = None,
    ) -> Text:
        """
        Get a prompt that is stored in Langfuse.

        :param name: Name of the prompt to retrieve.
        :param arguments: JSON object with prompt variables, e.g., '{"var": "value"}'.
        """
        try:
            args_dict = json.loads(arguments) if arguments else {}
            res = await get_prompt_handler(GetPrompt(name=name, arguments=args_dict))
            return Text(json.dumps(res.dict()))
        except Exception as e:
            await handle.error(f"Error: {e}")
            return Text("")

    return mcp
